### PR TITLE
Splitting diploma into training and prediction

### PR DESCRIPTION
**Pull request by romanie**

### CODE REVIEW

File: predict.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+if args.model == "logistic_regression":
+    model = linear_model.LogisticRegression(C=1, solver='lbfgs', 
+                                            multi_class='multinomial', max_iter=100, n_jobs=-1)
+elif args.model == "k_neighbors":
+    model = KNeighborsClassifier(n_neighbors=10, n_jobs=-1)
+else:
+    model = RandomForestClassifier(n_estimators=10, max_depth=20, n_jobs=-1)
+
+model_file = open('model.pkl', 'wb')
```
Comment (romanie) [inline predict.py]: Please configure model filename through flags so one could save different models.

File: predict.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+MAP_test = mean_average_precision(predicted_probability_distribution_test, 
+                                  model.classes_, 
+                                  test_y)
+print "MAP on test data: %.3f" % MAP_test
+MAP_train = mean_average_precision(predicted_probability_distribution_train, 
+                                   model.classes_, 
+                                   train_y) 
+print "MAP on train data: %.3f" % MAP_train
+
+# plot_learning_curve(model, args.model, train_X, train_y, ylim=None,
```
Comment (romanie) [inline predict.py]: Remove learning curve from here for now. All the code that you have wrote for the learning curve is saved in history on github, so don't worry about losing it.

File: predict.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        for i in xrange(0, len(row)):
+            if (row[i] > prob_of_target_class 
+                or (row[i] == prob_of_target_class and list_of_classes[i] > test_y[k])):
+                place += 1
+                if place >= 6: break
+        if place < 6:
+            total_error += 1.0 / place
+    return total_error / len(predicted_probability_distribution)
+
+def plot_learning_curve(estimator, title, X, y, ylim=None, cv=None,
```
Comment (romanie) [inline predict.py]: Please remove this learning curve code from here so this python script serve single purpose.

File: predict.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+parser = argparse.ArgumentParser()
+parser.add_argument("--model", choices=["random_forest", "k_neighbors", "logistic_regression"], default=["random_forest"], dest="model")
+parser.add_argument("--train_dataset_size", default=100000, type=int, dest="train_dataset_size")
+parser.add_argument("--test_dataset_size", default=0, type=int, dest="test_dataset_size")
+args = parser.parse_args()
+print "Model used in this run is %s" % colored(args.model, "red")
+print "Train dataset size is %s" % colored(args.train_dataset_size, "red")
+print "Test dataset size is %s" % colored(args.test_dataset_size, "red")
+
```
Comment (romanie) [inline predict.py]: too many non-useful new lines, please remove

File: train_model.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import time
+import argparse
+from sklearn import linear_model
+from sklearn.ensemble import RandomForestClassifier
+from sklearn.neighbors import KNeighborsClassifier
+import numpy as np
+from termcolor import colored
+import pickle
+
+def parse(line):
```
Comment (romanie) [inline train_model.py]: Please add a `common` module and move this function there.

File: train_model.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+start_time = time.time()
+
+parser = argparse.ArgumentParser()
+parser.add_argument("--model", choices=["random_forest", "k_neighbors", "logistic_regression"], default=["random_forest"], dest="model")
+parser.add_argument("--train_dataset_size", default=100000, type=int, dest="train_dataset_size")
+args = parser.parse_args()
+print "Model used in this run is %s" % colored(args.model, "red")
+print "Train dataset size is %s" % colored(args.train_dataset_size, "red")
+
+train = open ('train.csv', 'r')
```
Comment (romanie) [inline train_model.py]: Please wrap L28-L41 into a function 
`def load_dataset_from_file(filename, examples_count):`
and move it to the `common` module.

Yes, the function will return pair: `(train_X, train_y)`.

And yes, on the call-site it will look like

`train_X, train_y = load_dataset_from_file('train.csv', args.train_dataset_size)`

File: train_model.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+import numpy as np
+from termcolor import colored
+import pickle
+
+def parse(line):
+    tokens = line.split(',')
+    features = [tokens[3], tokens[4], tokens[5], tokens[13], tokens[14], tokens[15], tokens[16], 
+        tokens[17], tokens[20], tokens[21], tokens[22]]
+    features = [float (f) for f in features]
+    label = int (tokens[23])
```
Comment (romanie) [inline train_model.py]: Please add an optional argument specifying whether you are parsing labeled (train) data or you are parsing unlabeled (test) data.

`def parse(line, is_labeled=True):`

if this argument is `False` then you'll need to `return (features, None)` and avoid parsing the label. (`token[23]` is likely non-existent there)

Please also propagate it `load_dataset_from_file` function discussed in comment below so you have `def load_dataset_from_file(filename, examples_count, is_labeled=True):`. It should work well with both labeled and unlabeled data and reused between `train_model.py` and `predict_model.py`.

In `predict_model.py` you will be able to use it as

`test_X, _unused = load_dataset_from_file('test.csv', 50000000, false)`

File: train_model.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+if args.model == "logistic_regression":
+    model = linear_model.LogisticRegression(C=1, solver='lbfgs', 
+                                            multi_class='multinomial', max_iter=100, n_jobs=-1)
+elif args.model == "k_neighbors":
+    model = KNeighborsClassifier(n_neighbors=10, n_jobs=-1)
+else:
+    model = RandomForestClassifier(n_estimators=10, max_depth=20, n_jobs=-1)
+
+my_model = model.fit(train_X, train_y)
+model_file = open('model.pkl', 'wb')
```
Comment (romanie) [inline train_model.py]: Please make model filename configurable via flags.

Yes, this means just add a flag "--model_filename".

File: common.py [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+def parse(line, is_labeled=True):
+    tokens = line.split(',')
+    features = [tokens[3], tokens[4], tokens[5], tokens[13], tokens[14], tokens[15], tokens[16], 
+        tokens[17], tokens[20], tokens[21], tokens[22]]
+    features = [float (f) for f in features]
+    if is_labeled == False:
+    	return (features, None)
+    else: 
+    	label = int (tokens[23])
```
Comment (romanie) [inline common.py]: you can simplify and put here `return (features, int(tokens[23]))`

File: common.py [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+	    if cnt == 0:
+	        continue
+	    (features, label) = parse(line, is_labeled)
+	    if cnt <= examples_count:
+	        data_X.append(features)
+	        data_y.append(label)
+	    else: 
+	        break
+	if is_labeled == False:
+		return (data_X, None)
```
Comment (romanie) [inline common.py]: I think something wrong with indent here - remove extra spaces.

File: predict.py [hunk 11]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/python
+import time
+import numpy as np
+from termcolor import colored
+import pickle
+import common.py
+
+parser = argparse.ArgumentParser()
+parser.add_argument("--test_dataset_size", default=10000, type=int, dest="test_dataset_size")
```
Comment (romanie) [inline predict.py]: Make default 100 million, I believe most often you'll just want to predict for the whole test file.

File: predict.py [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+args = parser.parse_args()
+print "Test dataset size is %s" % colored(args.test_dataset_size, "red")
+
+test_X, test_y = load_dataset_from_file('test.csv', args.test_dataset_size, False)
+
+model_file = open('model.pkl', 'rb')
```
Comment (romanie) [inline predict.py]: Make model name configurable through flags.

File: predict.py [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+model = pickle.load(model_file)
+model_file.close()
+
+predicted_probability_distribution_test = model.predict_proba(test_X)
```
Comment (romanie) [inline predict.py:20]: You probably haven't finished writing this script. You need to output a file that will be readable by Kaggle.

File: common.py (hunk 9 above)
Comment (romanie) [inline common.py]: You can go even further and write `return (features, int(tokens[23]) if is_labeled else None)`.

File: common.py [hunk 14]
```diff
@@ -0,0 +1,32 @@
+import argparse
```
Comment (romanie) [inline common.py]: You don't need argparse here.

File: predict.py (hunk 11 above)
Comment (oksanaie) [inline predict.py]: I have just 700k test examples

File: train_model.py (hunk 8 above)
Comment (oksanaie) [inline train_model.py]: done

File: train_model.py (hunk 7 above)
Comment (oksanaie) [inline train_model.py]: done

File: predict.py [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+for row in predicted_probability_distribution:
+    k += 1
+    line = []
+    line.append(k)
+    classes_by_prob = []
+    for i in xrange(0, len(row)):
+        classes_by_prob.append([row[i], model.classes_[i]])
+        classes_by_prob.sort(reverse=True)
+    clusters =[]  
+    #clusters = [classes_by_prob[x][1] for x in xrange(0, 5)]
```
Comment (romanie) [inline predict.py]: Why have you commented this version? I think it is nicer than the one you currently have.

File: common.py (hunk 9 above)
Comment (oksanaie) [inline common.py]: done

File: common.py (hunk 10 above)
Comment (oksanaie) [inline common.py]: everything is ok

File: predict.py (hunk 12 above)
Comment (oksanaie) [inline predict.py]: done

File: predict.py (hunk 13 above)
Comment (oksanaie) [inline predict.py:20]: done

File: common.py (hunk 14 above)
Comment (oksanaie) [inline common.py]: done

File: predict.py [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+parser = argparse.ArgumentParser()
+parser.add_argument("--model_filename", dest="model_filename")
+parser.add_argument("--test_dataset_size", default=100000, type=int, dest="test_dataset_size")
+args = parser.parse_args()
+print "Test dataset size is %s" % colored(args.test_dataset_size, "red")
+
+test_X, _unused = load_dataset_from_file('test.csv', 700000, False)
+
+model_file = open('args.model_filename.pkl', 'rb')
```
Comment (romanie) [inline predict.py]: This is not magic, this is programming:
`open('args.model_filename.pkl', 'rb')` ->  `open(args.model_filename, 'rb')`

File: predict.py [hunk 17]
```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/python
+import time
+import numpy as np
+from termcolor import colored
+import pickle
+import common.py
+import csv
+
+parser = argparse.ArgumentParser()
+parser.add_argument("--model_filename", dest="model_filename")
```
Comment (romanie) [inline predict.py:12]: Make some reasonable default for model_filename flag, say "model.pkl".

File: common.py [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+            continue
+        (features, label) = parse(line, is_labeled)
+        if cnt <= examples_count:
+            data_X.append(features)
+            data_y.append(label)
+        else: 
+            break
+    if is_labeled == False:
+        return (data_X, None)
+    else:
```
Comment (romanie) [inline common.py]: Simplify this if to `return (features, int(tokens[23]) if is_labeled else None)`.

Comment (oksanaie) [inline common.py]: that was done
check that you opened new push

File: predict.py [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+test_X, _unused = load_dataset_from_file('test.csv', 700000, False)
+
+model_file = open('args.model_filename.pkl', 'rb')
+model = pickle.load(model_file)
+model_file.close()
+
+predicted_probability_distribution_test = model.predict_proba(test_X)
+
+result = []
+result.append([id,hotel_cluster])
```
Comment (romanie) [inline predict.py]: Space between `id` and `hotel_cluster`. Also rename `id` to `idx` since `id` is taken for a [builtin](https://docs.python.org/2/library/functions.html#id) function.

File: predict.py (hunk 15 above)
Comment (oksanaie) [inline predict.py]: I was not sure it is correct this way (syntax)

File: predict.py [hunk 20]
```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/python
+import time
+import numpy as np
+from termcolor import colored
+import pickle
+import common.py
```
Comment (romanie) [inline predict.py]: `common.py` -> `common`

File: predict.py [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+for row in predicted_probability_distribution:
+    k += 1
+    line = []
+    line.append(k)
+    classes_by_prob = []
+    for i in xrange(0, len(row)):
+        classes_by_prob.append([row[i], model.classes_[i]])
+        classes_by_prob.sort(reverse=True)
+    clusters = [classes_by_prob[x][1] for x in xrange(0, 5)]  
+    #clusters =[]
```
Comment (romanie) [inline predict.py]: Don't comment, remove unused code.

File: predict.py [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+k = -1
+for row in predicted_probability_distribution:
+    k += 1
+    line = []
+    line.append(k)
+    classes_by_prob = []
+    for i in xrange(0, len(row)):
+        classes_by_prob.append([row[i], model.classes_[i]])
+        classes_by_prob.sort(reverse=True)
+    clusters = [classes_by_prob[x][1] for x in xrange(0, 5)]
```
Comment (romanie) [inline predict.py]: You don't need a separate variable `clusters`, please inline this:
`line.append([classes_by_prob[x][1] for x in xrange(0, 5)])`

File: predict.py [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+import time
+import numpy as np
+from termcolor import colored
+import pickle
+import common
+import csv
+
+parser = argparse.ArgumentParser()
+parser.add_argument("--model_filename", default=model.pkl, dest="model_filename")
+#parser.add_argument("--test_dataset_size", default=700000, type=int, dest="test_dataset_size")
```
Comment (romanie) [inline predict.py]: Yes, I think you can remov this flag.

File: common.py [hunk 24]
```diff
@@ -0,0 +1,25 @@
+
```
Comment (romanie) [inline common.py]: Please remove L1 and L8.

File: common.py [hunk 25]
```diff
@@ -0,0 +1,25 @@
...
+        cnt += 1
+        if cnt == 0:
+            continue
+        (features, label) = parse(line, is_labeled)
+        if cnt <= examples_count:
+            data_X.append(features)
+            data_y.append(label)
+        else: 
+            break
+    return (features, int(tokens[23]) if is_labeled else None)
```
Comment (romanie) [inline common.py]: Add an empty new line in the end of the file.

File: train_model.py [hunk 26]
```diff
@@ -0,0 +1,38 @@
...
+print "Model used in this run is %s" % colored(args.model, "red")
+print "Train dataset size is %s" % colored(args.train_dataset_size, "red")
+
+train_X, train_y = load_dataset_from_file('train.csv', args.train_dataset_size, True)
+
+print "Input/Output time: %.3f" % (time.time() - start_time)
+train_start_time = time.time()
+
+if args.model == "logistic_regression":
+    model = linear_model.LogisticRegression(C=1, solver='lbfgs',
```
Comment (romanie) [inline train_model.py]: Nitpick: can you change your import from

`from sklearn import linear_model`

to

`from sklearn.linear_model import LogisticRegression` and use a simplified access here:

`model = LogisticRegression(...)` instead of `model = linear.LogisticRegression(...)`.

File: predict.py [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+    clusters = [classes_by_prob[x][1] for x in xrange(0, 5)]  
+    #clusters =[] 
+    # for x in xrange(0, 5):
+    #     clusters.append(classes_by_prob[x][1])
+    line.append(clusters)
+    result.append(line)
+
+with open("result.csv", "wb") as f:
+    writer = csv.writer(f)
+    writer.writerows(result)
```
Comment (romanie) [inline predict.py]: Add an empty new line in the end of the file.

**Comment by romanie** [comment]
Overall looks good to me - please address few nitpicks and run the submission to Kaggle. Then we can proceed with merging this in and celebrating tonight :)
